README.md
# lokka-transport-jwt-auth

A transport module for [Lokka](https://github.com/kadirahq/lokka) with JWT authentication. This is an extended version of [lokka-transport-http](https://github.com/kadirahq/lokka-transport-http) which uses JWT tokens. It also refreshes the token if it gets expired.

## Setup

```shell
npm i -S lokka lokka-transport-jwt-auth
```

## Usage

```js
import JWTAuthTransport from 'lokka-transport-jwt-auth';

// Write a function to create a new (JWT) token.
// It should return a token or a promise to a token.
// This function will be called again when the token
// is about to expire.
function refresh() {
  const token = 'get-jwt';
  return Promise.resolve(token);
}

// Create a new transport with an http endpoint and the refresh function
// If you need to pass additional options to the underlying HTTPTransport
// use this form: new JWTAuthTransport(endpoint, options, refresh);
const tr = new JWTAuthTransport('http://localhost:8090/path', refresh);

// You can start using the transport, requests will be processed
// as soon as the token is ready (using the refresh function).
tr.send(`{ latestPosts { title } }`).then(console.log);
```


CHANGELOG.md
# ChangeLog

### v1.3.2

* Remove jwt-js which introduce weird error related to crypto module

### v1.3.1

* Use jsonwebtoken as a proper dep

### v1.3.0

* Use jsonwebtoken again. We had some issues with the use of crypto in `jwt-js` with Meteor.

### v1.2.0

* Remove nofat

### v1.1.0

* Use jwt-js as the JWT implmentation.

### v1.0.0

* Initial release


src/__tests__/index.js
/* eslint-disable max-len */
const {describe, it, before, after} = global;
import sinon from 'sinon';
import {expect} from 'chai';
import {wink} from './sleep';
import {getToken} from './token';
import Transport from '../';

describe('Transport', function () {
  const globals = {
    refresh: () => getToken({time: Date.now()}),
  };

  before(function () {
    globals.clock = sinon.useFakeTimers(0);
  });

  after(function () {
    globals.clock.restore();
  });

  describe('constructor', function () {
    it('should initialize with 2 arguments', async function () {
      const tr = new Transport('endpoint', globals.refresh);
      expect(tr._endpoint).to.equal('endpoint');
      expect(tr._options).to.deep.equal({});
      expect(tr._refreshFn).to.equal(globals.refresh);
      expect(tr._transport).to.equal(null);

      // refresh
      await wink();
      const token = globals.refresh();
      expect(tr._transport._httpOptions.headers.Authorization).to.equal(`Bearer ${token}`);

      tr.close();
    });

    it('should initialize with 3 arguments', async function () {
      const tr = new Transport('endpoint', {foo: 'bar'}, globals.refresh);
      expect(tr._endpoint).to.equal('endpoint');
      expect(tr._options).to.deep.equal({foo: 'bar'});
      expect(tr._refreshFn).to.equal(globals.refresh);
      expect(tr._transport).to.equal(null);

      // refresh
      await wink();
      const token = globals.refresh();
      expect(tr._transport._httpOptions.headers.Authorization).to.equal(`Bearer ${token}`);

      tr.close();
    });

    it('should require the endpoint argument', function () {
      const fn = () => new Transport();
      expect(fn).to.throw('endpoint is required');
    });

    it('should require the refresh function argument', function () {
      const fn = () => new Transport('endpoint');
      expect(fn).to.throw('refresh function is required');
    });

    it('should not allow "Authorization" header', function () {
      const options = {headers: {Authorization: true}};
      const fn = () => new Transport('endpoint', options, globals.refresh);
      expect(fn).to.throw('the "Authorization" header should not exist');
    });

    it('should refresh the token 1 minute before it expires', async function () {
      const tr = new Transport('endpoint', globals.refresh);

      // refresh
      await wink();
      const token1 = globals.refresh();
      expect(tr._transport._httpOptions.headers.Authorization).to.equal(`Bearer ${token1}`);

      // advance the fake clock forward
      globals.clock.tick(1000 * 60 * 9);

      // refresh
      await wink();
      const token2 = globals.refresh();
      expect(tr._transport._httpOptions.headers.Authorization).to.equal(`Bearer ${token2}`);
    });
  });

  describe('send', function () {
    it('should send messages with transport', async function () {
      const tr = new Transport('endpoint', globals.refresh);

      // refresh
      await wink();
      tr._transport = {send: sinon.stub()};
      tr.send('{}', {}, 'op');
      expect(tr._transport.send.calledOnce).to.equal(true);
    });

    it('should not send if the transport is closed', function () {
      const tr = new Transport('endpoint', globals.refresh);
      const fn = () => tr.send('{}', {}, 'op');
      expect(fn).to.not.throw('transport is closed');

      tr.close();
      expect(fn).to.throw('transport is closed');
    });

    it('should add to waitlist if the transport is not ready', function () {
      const tr = new Transport('endpoint', Function());
      tr.send('{}', {}, 'op');
      expect(tr._waitlist.length).to.equal(1);
      tr.close();
    });

  });

  describe('close', function () {
    it('should stop the refresh loop', async function () {
      let calls = 0;
      function refresh() {
        calls++;
        return globals.refresh();
      }

      const tr = new Transport('endpoint', refresh);

      // refresh
      await wink();
      expect(calls).to.equal(1);

      for (let i = 2; i <= 5; ++i) {
        globals.clock.tick(1000 * 60 * 10);
        await wink();
        expect(calls).to.equal(i);
      }

      tr.close();
      globals.clock.tick(1000 * 60 * 10 * 10);
      await wink();
      expect(calls).to.equal(5);
    });
  });
});


src/__tests__/sleep.js
export function wink() {
  return new Promise(fn => process.nextTick(fn));
}

export function sleep(ms) {
  return new Promise(fn => setTimeout(fn, ms));
}


src/__tests__/token.js
import jwt from 'jsonwebtoken';

export function getToken(payload) {
  return jwt.sign(payload, 'secret', {expiresIn: '10m'});
}


src/index.js
import jwt from 'jsonwebtoken';
import HttpTransport from 'lokka-transport-http';

// if the token expires within the next N ms
// it will be refreshed immediately.
const MIN_REFRESH_TIMEOUT = 1000 * 60;

// if the token is not available, the job will
// wait for this number of milliseconds
const MAX_JOB_WAIT_TIME = 1000 * 10;
const ERR_JOB_TIMEOUT = new Error('job timeout');


export default class Transport {
  constructor(endpoint, arg2, arg3) {
    if (endpoint) {
      this._endpoint = endpoint;
    } else {
      throw new Error('endpoint is required');
    }

    // the options argument is optional
    if (typeof arg3 === 'function') {
      this._options = arg2;
      this._refreshFn = arg3;
    } else {
      this._options = {};
      this._refreshFn = arg2;
    }

    if (!this._refreshFn) {
      throw new Error('refresh function is required');
    }

    // the "Authorization" header will be used to store the JWT token
    // make sure the user is not using it for anything else.
    if (this._options.headers && this._options.headers.Authorization) {
      throw new Error('the "Authorization" header should not exist');
    }

    // The HTTP transport will be (re-)created with every token refresh.
    // This is done because we need to change headers with each refresh.
    this._transport = null;

    // queue requests here when this.token is null.
    // after refreshing the token, process the queue.
    this._waitlist = [];

    // true if the transport is manually closed
    // no further communication will be possible
    this._closed = false;

    // refresh immediately
    this._scheduleRefresh(0);
  }

  send(query, variables, opname) {
    if (this._closed) {
      throw new Error('transport is closed');
    }

    if (this._transport) {
      return this._transport.send(query, variables, opname);
    }

    return new Promise((resolve, reject) => {
      const job = {query, variables, opname, resolve, reject, done: false};
      this._waitlist.push(job);

      setTimeout(() => {
        if (!job.done) {
          job.done = true;
          reject(ERR_JOB_TIMEOUT);
        }
      }, MAX_JOB_WAIT_TIME);
    });
  }

  close() {
    this._transport = null;
    this._closed = true;
  }

  _processWaitlist() {
    const jobs = this._waitlist;
    this._waitlist = [];

    jobs.forEach(job => {
      const {query, variables, opname, resolve, reject, done} = job;
      if (!done) {
        job.done = true;
        this.send(query, variables, opname).then(resolve, reject);
      }
    });
  }

  async _refreshToken() {
    if (this._closed) {
      return;
    }

    try {
      const token = await this._refreshFn();
      if (!token) {
        throw new Error('invalid token');
      }

      const options = Object.assign({headers: {}}, this._options);
      options.headers.Authorization = `Bearer ${token}`;

      this._transport = new HttpTransport(this._endpoint, options);
      this._processWaitlist();

      // assuming the token has an expiration time
      // TODO handle tokens without expiration times
      const payload = jwt.decode(token);
      if (!payload || !payload.exp) {
        throw new Error('invalid token');
      }

      // schedule next token refresh
      const expires = payload.exp * 1000;
      this._scheduleRefresh(expires);
    } catch (e) {
      // console.log the error??
      this._transport = null;
      this._scheduleRefresh(0);
    }
  }

  _scheduleRefresh(expires) {
    const now = Date.now();
    const timeLeft = expires - now;

    if (timeLeft <= MIN_REFRESH_TIMEOUT) {
      this._refreshToken();
      return;
    }

    // add some slack time to avoid queuing
    const timeout = timeLeft - MIN_REFRESH_TIMEOUT;
    setTimeout(() => this._refreshToken(), timeout);
  }
}
